File: Data Structures & Algorithms/duplicate-integer/submission-3.cpp
class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
        unordered_map<int, int> freq;

        for (auto i : nums) {
            if (freq[i]) {
                return true;
            }
            freq[i]++;
        }
        return false;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1]
        suffix = [1]
        run = 1
        for i in range(1, len(nums)):
            run *= nums[i-1]
            prefix.append(run)
        rev = nums[::-1]
        run = 1
        for i in range(1, len(rev)):
            run *= rev[i-1]
            suffix.append(run)
        suffix = suffix[::-1]
        res = []
        
        for i in range(len(nums)):
            res.append(prefix[i] * suffix[i])
        
        return res